### DISCUSSION TITLE
Problems with PersistJobDataAfterExecution

### DISCUSSION THREAD

**Discussion by asherber** [quartznet/quartznet#1334] [Q&A]
I'm having a problem with my first job with `PersistJobDataAfterExecution`. I'm using a clustered scheduler against a SQL Server db, and I've got this job configured to run every 5 seconds:

```csharp
[DisallowConcurrentExecution]
[PersistJobDataAfterExecution]
public class TestJob : IJob
{
    public Task Execute(IJobExecutionContext context)
    {
        var current = DateTime.Now;
        var last = context.MergedJobDataMap.GetDateTimeValue("LastRun");
        Console.WriteLine($"\n*** Job fired at {current}");
        Console.WriteLine($"*** Last run was {last}\n");
        context.JobDetail.JobDataMap.Put("LastRun", current.ToString());
        return Task.CompletedTask;
    }
}
```

When I schedule and run this job, it runs only once, and the updated data is not written to the job's data map. 

If I remove the `DisallowConcurrentExecution` attribute, then the job does repeat every 5 seconds, as expected, but it still doesn't store the updated data.

(If I schedule this same job with a `RAMJobStore`, then it behaves as expected, with or without the `DisallowConcurrentExecution` attribute.)

I'm attaching two logs, one showing what happens with both attributes, and one showing what happens just with `PersistJobDataAfterExecution`. The lines starting with *** are written by my application; the rest is debug logging from Quartz. 

[Persist and Disallow.txt](https://github.com/quartznet/quartznet/files/7325922/Persist.and.Disallow.txt)
[Persist only.txt](https://github.com/quartznet/quartznet/files/7325923/Persist.only.txt)

Can someone tell me what I'm missing here?

**Comment by lahma** [comment]
With `RamJobStore` you will observe a "working version" as it's in memory and the attribute is not strictly required. Quartz __should__ write the state to database if the following are true: `jobDetail.PersistJobDataAfterExecution` and `jobDetail.JobDataMap.Dirty`: https://github.com/quartznet/quartznet/blob/2a679fa0562d871984f8cdac4b069cd398305b9e/src/Quartz/Impl/AdoJobStore/JobStoreSupport.cs#L3019-L3041

Maybe you could debug against Quartz sources and see what is actually going on there? It's almost like the attribute is not present on the running application.

**Reply by asherber** [reply]
@lahma The issues with persisting job data have to do with the combination of `BinaryObjectSerializer` and `UseProperties = "true"`. With this combination, `StdAdoDelegate.SerializeJobData()` calls `SerializeProperties()`. This turns the map into a `NameValueCollection`, and there's a serialization error because that class is not marked as `Serializable`. If `UseProperties == "false"`, or if I use `JsonObjectSerializer` with either value of `UseProperties`, everything works as expected.

I don't know whether this is intentional, but it strikes me as unexpected. It might be nice if `BinaryObjectSerializer` ignored the value of `UseProperties`, but I don't know what other problems that might cause.

As far as why the serialization error prevented the job from recurring if `DisallowConcurrentExecution` is set, I didn't dig deep enough to get an answer -- once the serialization issue was fixed, the job recurred regardless of whether that attribute was present. But that also seems to be problematic, since the serializer could throw an exception for any number of reasons. I think what may happen is that when [this line](https://github.com/quartznet/quartznet/blob/2a679fa0562d871984f8cdac4b069cd398305b9e/src/Quartz/Impl/AdoJobStore/JobStoreSupport.cs#L3025) throws an exception, the call to `DeleteFiredTrigger` a few lines later never happens. This leaves an entry in `qrtz_fired_triggers` with a status of `EXECUTING`, which blocks the job from recurring in the future (because Quartz thinks it's already running). I'm not sure that's desirable either -- at a glance, it looks like that last `try..catch` block should itself be in a `finally` for the preceding block, to make sure it gets called.

**Reply by lahma** [reply]
Hmm... Looking at the code you are just putting string key-value-pairs in which should be fine. Generally you should favor putting string keys and values and using JSON serializer if possible for new setups. I didn't see any serialization errors in the provided logs though.

**Reply by asherber** [reply]
Yeah, I'm not sure why those didn't get logged. But when I added an NUnit test and stepped through it, I could see it happen. I can provide the test file if you want.

**Reply by lahma** [reply]
Ideally if you can track why those don't get logged and provide a PR to fix things, that would be great. These are silent errors are really nasty when there's no way to know what caused the error.
